Add code sample for `annotated_class_clean_coding` (#13)
wangdi4/oneAPI-samples: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/annotated_class_clean_coding/README.md

# `Annotated Classes Clean Coding` Sample

> This Intel® FPGA tutorial demonstrates how to write cleaner code by: (i) using `properties_t` class to simplify the declaration of annotated types, and (ii) using the `alloc_annotated` helper function to create annotated pointers more neatly than the standard USM allocation APIs.

Reviewer: ```suggestion
This Intel® FPGA tutorial demonstrates how to use the included `annotated_class_util.hpp` header file to write cleaner code. This header lets you: 
* use the `properties_t` class to simplify the declaration of annotated types
* use the `alloc_annotated` helper function to create annotated pointers more neatly than the standard USM allocation APIs.
```

```mermaid
flowchart LR
   tier1("Tier 1: Get Started")
   tier2("Tier 2: Explore the Fundamentals")
   tier3("Tier 3: Explore the Advanced Techniques")
   tier4("Tier 4: Explore the Reference Designs")

   tier1 --> tier2 --> tier3 --> tier4

   style tier1 fill:#f96,stroke:#333,stroke-width:1px,color:#fff
   style tier2 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
   style tier3 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
   style tier4 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
```

Find more information about how to navigate this part of the code samples in the [FPGA top-level README.md](/DirectProgramming/C++SYCL_FPGA/README.md).
You can also find more information about [troubleshooting build errors](/DirectProgramming/C++SYCL_FPGA/README.md#troubleshooting), [running the sample on the Intel® DevCloud](/DirectProgramming/C++SYCL_FPGA/README.md#build-and-run-the-samples-on-intel-devcloud-optional), [using Visual Studio Code with the code samples](/DirectProgramming/C++SYCL_FPGA/README.md#use-visual-studio-code-vs-code-optional), [links to selected documentation](/DirectProgramming/C++SYCL_FPGA/README.md#documentation), etc.

## Purpose

> The `hls_flow_interfaces/mmhost` code sample demonstrates the usage of `annotated_arg` class in customizing Avalon memory-mapped interfaces for the FPGA IP component. (To learn more about `annotated_arg` class, refer to the code sample [hls_flow_interfaces](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/hls_flow_interfaces)). This tutorial will further demonstrate how to utilize certain helper function/class to write cleaner designs with `annotated_arg`.

Reviewer: ```suggestion
The `annotated_arg` class can be used to customize Avalon memory-mapped interfaces for FPGA IP components. (To learn more about `annotated_arg` class, refer to the code sample [mmhost](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/hls_flow_interfaces/mmhost)). This tutorial will further demonstrate how to use the helper code in `annotated_class_util.hpp` to clean up your code that uses `annotated_arg`. You may add `annotated_class_util.hpp` to your designs.
```

### Use helper class `properties_t` to simplify the declaration of annotated classes

> The code sample [hls_flow_interfaces/mmhost](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/hls_flow_interfaces/mmhost) has demonstrated how to declare an annotated_arg type using `decltype`. For example,

Reviewer: ```suggestion
Most oneAPI code assigns properties to an `annotated_arg` type using `decltype`. For example,
```

You have referred to this sample 3 times now. I've deleted all but one; we don't need to link to the same thing 3 times :)

```c++
using ArgProps = decltype(sycl::ext::oneapi::experimental::properties{
                              propertiesA,
                              propertiesB,
                              propertiesC});

> sycl::ext::oneapi::experimental::annotated_arg<int *, ArgProps> arg_x;

Reviewer: ```suggestion
using ArgProps = decltype(sycl::ext::oneapi::experimental::properties{
                              propertiesA,
                              propertiesB,
                              propertiesC});
sycl::ext::oneapi::experimental::annotated_arg<int *, ArgProps> arg_x;
```

```

> The header file "annotated_class_util.hpp" provides a helper class `properties_t` (in `fpga_tools` namespace) that simplifies the code as follows

Reviewer: ```suggestion
The header file "annotated_class_util.hpp" provides a helper class `properties_t` (in `fpga_tools` namespace) that lets you simplify your code as follows:
```

```c++
#include "annotated_class_util.hpp"

using ArgProps = fpga_tools::properties_t<propertiesA, propertiesB, propertiesC>;

> sycl::ext::oneapi::experimental::annotated_arg<int *, ArgProps> arg_x;

Reviewer: ```suggestion
#include "annotated_class_util.hpp"
<...>
using ArgProps = fpga_tools::properties_t<propertiesA, 
                                                                      propertiesB, 
                                                                      propertiesC>;
sycl::ext::oneapi::experimental::annotated_arg<int *, ArgProps> arg_x;
```

please align it correctly :)

Author: I changed the placeholders "propertiesA/B/C" which could be confusing to concrete properties..

```

> > **Note**: to use the helper class `properties_t`, you need to add the compiler flag "-std=c++20" in your compiling command. "-std=c++20" has been added to the CMakeLists.txt file of this tutorial.

Reviewer: ```suggestion
> **Note**: to use the helper class `properties_t`, you need to add the compiler flag "-std=c++20" to your compilation command. "-std=c++20" has been added to the `USER_FLAGS` variable in the CMakeLists.txt file of this tutorial:
> ```
> set(USER_FLAGS ${USER_FLAGS};-std=c++20)
>```
```

> > **Note**: The helper class `properties_t` can be used on other annotated classes that specify properties in form of `decltype(sycl::ext::oneapi::experimental::properties{...})`, such as [device_global](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/device_global) and [pipe](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/hostpipes), provided the header file "annotation_class_util.hpp".

Reviewer: ```suggestion
> **Note**: The helper class `properties_t` can be used on other annotated classes that specify properties in form of `decltype(sycl::ext::oneapi::experimental::properties{...})`, such as [device_global](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/device_global) and [pipe](DirectProgramming/C++SYCL_FPGA/Tutorials/Features/hls_flow_interfaces/streaming_data_interfaces), provided the header file "annotation_class_util.hpp".
```
I'd prefer you link to the pipe tutorial that demonstrates pipes for SYCL HLS components rather than the host pipes tutorial, which is pretty much useless without an appropriate BSP :)

### Use helper function `alloc_annotated` to allocate host/shared memory with properties that match with Avalon memory-mapped host interfaces

> When testing the functionality and performance of an IP component using FPGA simulation, it is the responsibility of the SYCL host code to provide valid runtime values that matches with the properties specified in the `annotated_arg` kernel arguments. This section shows how to use the helper function `alloc_annotated` (defined in the utility header file "annotation_class_util.hpp") to ensure such requirement. 

Reviewer: ```suggestion
While SYCL device code must specify `annotated_arg` properties at **compile** time, the SYCL specification allows host code to specify these properties at **run** time. If you are designing and testing an IP component, this additional flexibility is of no use to you, so you can use the helper function `alloc_annotated()` (defined in the utility header file "annotation_class_util.hpp") to simplify your code, while still ensuring that the properties in your host allocations match the properties you defined in your device code. 
```

Author: updated

> In the example, the device code defines a SYCL kernel functor that multiplies a factor of 2 to an input vector as follows

Reviewer: ```suggestion
In the example, the device code defines a SYCL kernel functor that multiplies a factor of 2 to an input vector as follows:
```

# `Annotated Classes Clean Coding` Sample

> This Intel® FPGA tutorial demonstrates how to use the included `annotated_class_util.hpp` header file to write cleaner code. This header lets you:

Reviewer 2: Introduce lists with a full sentence. For example, you could say, "You can use this header file to write cleaner code as follows:"

Author: Updated.

> * use the `properties_t` class to simplify the declaration of annotated types

Reviewer 3: ```suggestion
* use the `properties_t` alias to simplify the declaration of annotated types with groups of properties
```
Update other uses similarly

> * use the `properties_t` alias to simplify the declaration of annotated types with groups of properties

Reviewer 2: Rewrite to emphasize the benefit and not the use:

Simply the declaration of...by using the `properties_t` alias.

Author: Updated.

> * use the `alloc_annotated` helper function to create annotated pointers more neatly than the standard USM allocation APIs.

Reviewer 2: Create annotated pointers more neatly....by using the `alloc_annotated` helper function.

Author: Updated.

| Hardware                          | Intel® Agilex® 7, Arria® 10, Stratix® 10, and Cyclone® V FPGAs
| Software                          | Intel® oneAPI DPC++/C++ Compiler
| What you will learn               | How to use the provided helper funtion to allocate host/shared memory with properties for your FPGA IP components
| Time to complete                  | 10 minutes

> **Note**: Even though the Intel DPC++/C++ OneAPI compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.
>
> To use the simulator flow, Intel® Quartus® Prime Pro Edition (or Standard Edition when targeting Cyclone® V) and one of the following simulators must be installed and accessible through your PATH:
> - Questa*-Intel® FPGA Edition
> - Questa*-Intel® FPGA Starter Edition
> - ModelSim® SE
>
> When using the hardware compile flow, Intel® Quartus® Prime Pro Edition (or Standard Edition when targeting Cyclone® V) must be installed and accessible through your PATH.
>
> :warning: Make sure you add the device files associated with the FPGA that you are targeting to your Intel® Quartus® Prime installation.

> > **Note**: This tutorial will not work for a Full System compile as it demonstrates a SYCL HLS flow specific feature.

Reviewer 2: "Full System" is now "FPGA acceleration".

This tutorial does not apply to the FPGA acceleration flow. It demonstrates features specific to the SYCL HLS flow.

Author: Updated.

```

Find more information about how to navigate this part of the code samples in the [FPGA top-level README.md](/DirectProgramming/C++SYCL_FPGA/README.md).
You can also find more information about [troubleshooting build errors](/DirectProgramming/C++SYCL_FPGA/README.md#troubleshooting), [running the sample on the Intel® DevCloud](/DirectProgramming/C++SYCL_FPGA/README.md#build-and-run-the-samples-on-intel-devcloud-optional), [using Visual Studio Code with the code samples](/DirectProgramming/C++SYCL_FPGA/README.md#use-visual-studio-code-vs-code-optional), [links to selected documentation](/DirectProgramming/C++SYCL_FPGA/README.md#documentation), etc.

## Purpose
The `annotated_arg` class can be used to customize Avalon memory-mapped interfaces for FPGA IP components. (To learn more about the `annotated_arg` class, refer to the code sample [mmhost](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/hls_flow_interfaces/mmhost)). This tutorial will further demonstrate how to use the helper code in `annotated_class_util.hpp` to clean up your code that uses `annotated_arg`. You may add `annotated_class_util.hpp` to your designs.

> ### Use `properties_t` to simplify the declaration of annotated types with groups of properties

Reviewer 2: Simply the declaration....

Also, headings should use headline-style capitalization:

Simplify the Declaration of Annotated Types By Using The `properties_t` Alias? Template Class? with Groups of Properties

Author: Updated.

      sycl::ext::intel::experimental::buffer_location<1>,
      sycl::ext::intel::experimental::awidth<32>,
      sycl::ext::intel::experimental::dwidth<256>,
      sycl::ext::oneapi::experimental::alignment<32>,
      sycl::ext::intel::experimental::maxburst<8>,
      sycl::ext::intel::experimental::latency<0>>>;
sycl::ext::oneapi::experimental::annotated_arg<int *, ArgProps> arg_x;
```

> **Note**: to use the `properties_t` alias, you need to add the compiler flag `-std=c++20` to your compilation command. `-std=c++20` has been added to the `USER_FLAGS` variable in the CMakeLists.txt file of this tutorial:
> ```
> set(USER_FLAGS ${USER_FLAGS} -std=c++20)
>```

> **Note**: The `properties_t` alias can be used on other annotated classes that specify properties in form of `decltype(sycl::ext::oneapi::experimental::properties{...})`, such as [device_global](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/device_global) and [pipe](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/hls_flow_interfaces/streaming_data_interfaces), assuming you have included the header file "annotation_class_util.hpp" that appears in this code sample.

> ### Use helper function `alloc_annotated` to allocate host/shared memory with properties that match with Avalon memory-mapped host interfaces

Reviewer 2: Allocate Host/Shared Memory....

Author: Updated.

While SYCL device code must specify `annotated_arg` properties at **compile** time, the SYCL specification allows host code to specify these properties at **run** time. If you are designing and testing an IP component, this additional flexibility is of no use to you, so you can use the helper function `alloc_annotated()` (defined in the utility header file "annotation_class_util.hpp") to simplify your code, while still ensuring that the properties in your host allocations match the properties you defined in your device code.

In the example, the device code defines a SYCL kernel functor that multiplies a factor of 2 to an input vector as follows:

```c++
using annotated_arg_t= sycl::ext::oneapi::experimental::annotated_arg<
      int *, fpga_tools::properties_t<
            sycl::ext::intel::experimental::buffer_location<0>,
            sycl::ext::oneapi::experimental::alignment<256>>>;

struct MyIP {
  annotated_arg_t a;

  int size;

  void operator()() const {
   #pragma unroll 8
    for (int i = 0; i < size; i++) {
      a[i] *= 2;
    }
  }
};
```

> The type of the kernel argument `a` is annotated_arg specified with properties `buffer location` and `alignment`. To test the IP component in the FPGA simulation, the SYCL host code is required to provide for the kernel arguments runtime values that match with the `buffer location` and `alignment`. The helper function `alloc_annotated` (in `fpga_tools` namespace) provides this functionality with code simplicity, as it takes the annotated_arg type as the only template parameter and returns an instance of the same annotated_arg type, as the following code snippet shows

Reviewer 2: annotated_arg -> `annotated_arg`

Author: Updated.

```c++
#include "annotated_class_util.hpp"

constexpr int kN = 8;
annotated_arg_t array_a = fpga_tools::alloc_annotated<annotated_arg_t>(kN, q);

// Initialize the host array
...
// Launch the kernel
q.single_task(MyIP{array_a, kN}).wait();
```

> > **Note**: "alloc_annotated" function only takes `annotated_arg` over **pointer types** as the valid template parameters.

Reviewer 3: ```suggestion
> **Note**: "alloc_annotated" function only takes `annotated_arg` templated with **pointer types** as the valid template parameter.
```

## Purpose

> The `annotated_arg` class can be used to customize Avalon memory-mapped interfaces for FPGA IP components. (To learn more about `annotated_arg` class, refer to the code sample [mmhost](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/hls_flow_interfaces/mmhost)). This tutorial will further demonstrate how to use the helper code in `annotated_class_util.hpp` to clean up your code that uses `annotated_arg`. You may add `annotated_class_util.hpp` to your designs.

Reviewer: ```suggestion
The `annotated_arg` class can be used to customize Avalon memory-mapped interfaces for FPGA IP components. (To learn more about the `annotated_arg` class, refer to the code sample [mmhost](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/hls_flow_interfaces/mmhost)). This tutorial will further demonstrate how to use the helper code in `annotated_class_util.hpp` to clean up your code that uses `annotated_arg`. You may add `annotated_class_util.hpp` to your designs.
```

Most oneAPI code assigns properties to an `annotated_arg` type using `decltype`. For example,
```c++
using ArgProps = decltype(sycl::ext::oneapi::experimental::properties{
      sycl::ext::intel::experimental::buffer_location<1>,
      sycl::ext::intel::experimental::awidth<32>,
      sycl::ext::intel::experimental::dwidth<256>,
      sycl::ext::oneapi::experimental::alignment<32>,
      sycl::ext::intel::experimental::maxburst<8>,
      sycl::ext::intel::experimental::latency<0>});
sycl::ext::oneapi::experimental::annotated_arg<int *, ArgProps> arg_x;
```

The header file "annotated_class_util.hpp" provides a helper class `properties_t` (in `fpga_tools` namespace) that lets you simplify your code as follows:
```c++
#include "annotated_class_util.hpp"

using ArgProps = fpga_tools::properties_t<
      sycl::ext::intel::experimental::buffer_location<1>,
      sycl::ext::intel::experimental::awidth<32>,
      sycl::ext::intel::experimental::dwidth<256>,
      sycl::ext::oneapi::experimental::alignment<32>,
      sycl::ext::intel::experimental::maxburst<8>,
      sycl::ext::intel::experimental::latency<0>>>;
sycl::ext::oneapi::experimental::annotated_arg<int *, ArgProps> arg_x;
```

> > **Note**: to use the helper class `properties_t`, you need to add the compiler flag "-std=c++20" to your compilation command. "-std=c++20" has been added to the `USER_FLAGS` variable in the CMakeLists.txt file of this tutorial:

Reviewer: ```suggestion
> **Note**: to use the helper class `properties_t`, you need to add the compiler flag `-std=c++20` to your compilation command. `-std=c++20` has been added to the `USER_FLAGS` variable in the CMakeLists.txt file of this tutorial:
```

> ```
> set(USER_FLAGS ${USER_FLAGS} -std=c++20)
>```

> > **Note**: The helper class `properties_t` can be used on other annotated classes that specify properties in form of `decltype(sycl::ext::oneapi::experimental::properties{...})`, such as [device_global](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/device_global) and [pipe](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/hls_flow_interfaces/streaming_data_interfaces), provided the header file "annotation_class_util.hpp".

Reviewer: ```suggestion
> **Note**: The `properties_t` alias can be used on other annotated classes that specify properties in form of `decltype(sycl::ext::oneapi::experimental::properties{...})`, such as [device_global](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/device_global) and [pipe](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/hls_flow_interfaces/streaming_data_interfaces), assuming you have included the header file "annotation_class_util.hpp" that appears in this code sample.
```
